Cargo.toml

```diff
@@ -15,6 +15,7 @@ async = ["flume/async"]
 default = ["async"]
 
 [dependencies]
-flume = { version = "0.10", default-features = false } # channel between threads
-log = "0.4.14"                                         # logging
-nix = "0.23.1"                                         # socket APIs
+flume = { version = "0.10", default-features = false }     # channel between threads
+log = "0.4.14"                                             # logging
+socket2 = { version = "0.3.19", features = ["reuseport"] } # socket APIs
```

Comment:
Unfortunately, `socket2` seems to split into two branches now:  `0.3` and `0.4`.  We opt to use `0.3` for now mainly due to its socket [`recv_from`](https://docs.rs/socket2/0.3.19/socket2/struct.Socket.html#method.recv_from) method .

Reply:
There would be a `Read` trait implementation (for `read`) which we could adapt: https://github.com/rust-lang/socket2/blob/master/src/socket.rs#L1816 if we wanted to go with `socket2` `0.4` support.

Reply:
`read` of `Read` trait calls `recv`, hence returns only the buffer, not the source address.  Currently we are using `recv_from` which returns both the buffer and the source address.  It's true that we are not utilizing the source address at the moment (besides logging) as we only do multicast, but I'm not sure if we want it for future or not. It will be a trade-off.

Reply:
I am sorry for the delay, kind of... forgot, what I mean is that we could use the function that already does it via the trait and adapt it.

```rust
fn read_from(sock: &mut Socket), buf: &mut [u8]) -> IpAddr {
        let buf = unsafe { &mut *(buf as *mut [u8] as *mut [MaybeUninit<u8>]) };
        self.recv_from(buf)
}
```

This is more pseudo-code than anything.

Reply:
no problem. I didn't want to use `unsafe` in this code base though. In comparison,  changing `recv_from` to `read` seems to strike a better trade-off.

Reply:
I finally moved to the current `socket2` release (i.e. 0.4 branch) to be future-proof. I felt it's the better trade-off.  And I plan to merge these changes if no other issues.